### PR TITLE
Expand is_functional_schema to work with torch._C._FunctionSchema

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #123454
* #123453
* #123437
* #123129
* #123110
* #123109
* __->__ #123108

Previously it worked with torchgen.model.FunctionSchema. This PR extends
it to work with torch._C._FunctionSchema by making
torchgen.model.FunctionSchema look more like torch._C._FunctionSchema.

Test Plan:
- new tests

### CODE REVIEW

File: torch/_library/utils.py [hunk 1]
```diff
@@ -78,24 +78,29 @@ def is_functional_schema(schema: Any) -> bool:
...
-    rets = schema.returns
-    is_non_mutating_view = len(rets) > 0 and any(
-        r.annotation is not None and not r.annotation.is_write for r in rets
-    )
-    if is_non_mutating_view:
-        return False
-    if not schema.returns:
-        return False
-    return True
+    return is_functional(schema)
```
Comment (albanD) [inline torch/_library/utils.py:103]: nit: this can be re-ordered up by doing a single assert + if + this call.

File: torch/_C/__init__.pyi.in [hunk 2]
```diff
@@ -854,6 +854,7 @@ class FunctionSchema:
     returns: List[Argument]
     name: str
     overload_name: str
+    is_mutable: _bool
```
Comment (albanD) [inline torch/_C/__init__.pyi.in:857]: Why do we have this field and not just look it up on the arguments?

Comment (albanD) [reply torch/_C/__init__.pyi.in:857]: Isn't that what you do in your function is_functional that you added below?

Comment (zou3519) [reply torch/_C/__init__.pyi.in:857]: Is your suggestion to delete `is_mutable` and add a helper function `is_mutable` to do what it previously did? If so, there already is an `is_mutable` helper function in C++ (that's what this field is).

NB: `is_mutable` and `is_functional_schema` below aren't opposites :/ we have too many variants of these checks flying around and I'm trying to not break any existing code (is_mutable exists already today).

Comment (zou3519) [reply torch/_C/__init__.pyi.in:857]: NB: is_mutable is a @property, so we're good.

File: torch/_library/utils.py [hunk 3]
```diff
@@ -78,24 +78,29 @@ def is_functional_schema(schema: Any) -> bool:
     - it has at least one return
     """
 
+    def is_functional(schema):
+        if schema.is_mutable:
+            return False
+        rets = schema.returns
+        is_non_mutating_view = len(rets) > 0 and any(
+            r.alias_info is not None and not r.alias_info.is_write for r in rets
```
Comment (albanD) [inline torch/_library/utils.py:86]: How could "is_write" be true here since we check for is_mutable False above?

File: torch/_library/utils.py [hunk 4]
```diff
@@ -78,24 +78,29 @@ def is_functional_schema(schema: Any) -> bool:
...
+        )
+        if is_non_mutating_view:
+            return False
+        if not schema.returns:
+            return False
```
Comment (albanD) [inline torch/_library/utils.py:91]: this is a weird schema: not inplace and no output haha

**Review by albanD — APPROVED** [review]
Only nits in the logic that is weird, but most likely due to the complexity of the schema so not blocking the PR

File: torch/_library/utils.py (hunk 3 above)
Comment (zou3519) [reply torch/_library/utils.py:86]: is_mutable only checks that the inputs are written to :/
All this code is pre-existing, I am not changing the logic. I agree that it's messy.

File: torch/_library/utils.py (hunk 4 above)
Comment (zou3519) [reply torch/_library/utils.py:91]: this usually just means the user did write an inplace op but forgot to label it as inplace.

**Pull request closed without merging**